src/theory/quantifiers/sygus/sygus_reconstruct.h

```diff
@@ -0,0 +1,300 @@
+/*********************                                                        */
+/*! \file sygus_reconstruct.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Abdalrhman Mohamed
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief utility for reconstructing terms to match a grammar
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__THEORY__QUANTIFIERS__SYGUS_RECONSTRUCT_H
+#define CVC4__THEORY__QUANTIFIERS__SYGUS_RECONSTRUCT_H
+
+#include <map>
+#include <vector>
+
+#include "theory/quantifiers/sygus/rcons_obligation_info.h"
+#include "theory/quantifiers/sygus/rcons_type_info.h"
+
+namespace CVC4 {
+namespace theory {
+namespace quantifiers {
+
+using ObligationSet = std::unordered_set<Node, NodeHashFunction>;
+using TypeObligationSetMap =
+    std::unordered_map<TypeNode, ObligationSet, TypeNodeHashFunction>;
+
+/** SygusReconstruct
+ *
+ * This class implements an algorithm for reconstructing a given term such that
+ * the reconstructed term is equivalent to the original term and its syntax
+ * matches a specified grammar.
+ *
+ * Goal: find a term g in sygus type T that is equivalent to builtin term t.
+ *
+ * rcons(t_0, T_0) returns g
+ * {
+ *   Obs: A map from sygus types T to a set of triples to reconstruct into T,
+ *        where each triple is of the form (k, t, s), where k is a skolem of
+ *        type T, t is a builtin term of the type encoded by T, and s is a
+ *        possibly null sygus term of type T representing the solution.
+ *
+ *   Sol: A map from a skolems k to slutions of their corresponding obligations
+ *        such that Obs[T] contains entries (k, t, Sol[k]).
+ *
+ *   CandSols : A map from a skolem k to a set of possible solutions for its
+ *              corresponding obligation. Whenever there is a successful match,
+ *              it is added to this set.
+ *
+ *   Pool : A map from a sygus type T to a set of enumerated terms in T.
+ *          The terms in this pool are patterns whose builtin analogs are used
+ *          for matching against the terms to reconstruct t in (k, t, s).
+ *
+ *   let k_0 be a fresh skolem of sygus type T_0
+ *   Obs[T_0] += (k_0, t_0, null)
+ *
+ *   while Sol[k_0] == null
+ *     // enumeration phase
+ *     for each subfield type T of T_0
+ *       s[z] = nextEnum(T_{z})
+ *       if (s[z] is ground)
```

Comment:
same point as above about "ground"

Reply:
This one is a bit different. The current implementation of `nextEnum` uses variables, not skolems. I don't know whether or not the distinction matters in this context. If you still think "pure" is more appropriate, I will go ahead and make the change.

Reply:
I see, ground is better indeed. However I think this is still a bit confusing. The variables `z` are ranging over what?

Reply:
They variables `z` range over all terms of their type. By their type, I mean subfield types of `T_0`. I modified the comment. Please take another look and let me know if it's still unclear.